x/msg_delegation/spec/msg_delegation.md

```diff
@@ -0,0 +1,125 @@
+# `msg_delegation` Module
+
+The `msg_delegation` module provides support for granting arbitrary capabilities
+from one account (the granter) to another account (the grantee). Capabilities
+must be granted for a particular type of `sdk.Msg` one by one using an implementation
+of `Capability`.
+
+## Types
+
+Capabilities determine exactly what action is delegated. They are extensible
+and can be defined for any sdk.Msg type even outside of the module where the Msg is defined.
+
+### Capability
+
+```go
+type Capability interface {
+	// MsgType returns the type of Msg's that this capability can accept
+	MsgType() sdk.Msg
+	// Accept determines whether this grant allows the provided action, and if
+	// so provides an upgraded capability grant
+	Accept(msg sdk.Msg, block abci.Header) (allow bool, updated Capability, delete bool)
+}
+```
+
+For example a `SendCapability` like this is defined for `MsgSend` that takes
+a `SpendLimit` and updates it down to zero:
+
+```go
+type SendCapability struct {
+	// SpendLimit specifies the maximum amount of tokens that can be spent
+	// by this capability and will be updated as tokens are spent. If it is
+	// empty, there is no spend limit and any amount of coins can be spent.
+	SpendLimit sdk.Coins
+}
+
+func (cap SendCapability) MsgType() sdk.Msg {
+	return bank.MsgSend{}
+}
+
+func (cap SendCapability) Accept(msg sdk.Msg, block abci.Header) (allow bool, updated Capability, delete bool) {
+	switch msg := msg.(type) {
+	case bank.MsgSend:
+		left, invalid := cap.SpendLimit.SafeSub(msg.Amount)
+		if invalid {
+			return false, nil, false
+		}
+		if left.IsZero() {
+			return true, nil, true
+		}
+		return true, SendCapability{SpendLimit: left}, false
```

Comment:
pp: I assume the flow with `MsgExecDelegated ` is like this:
```go
for _, msg := range tx.Msgs{
  c, expiry := store.GetCapability(ctx, tx.Grantee, msg.GetSigners()[0],msg)
  if c ==nil { continue // or return not found error? }
  if expiry <=now {  continue 
  // or return error? not found or expired?
  // what todo with c on in the storage? if we delete it now, then "expired" is not a good error msg. 
  }
  // when all good
  ok, newC, delete := c.Accept(msg, block)
  if !ok { continue // or error? only a generic error would be possible}
  if delete{
    _ = store.Delete(ctx, ctx, tx.Grantee, msg.GetSigners()[0],msg)
  } else {
  _ = store.Save(ctx, tx.Grantee, msg.GetSigners()[0],newC, expiry)
 }
}
```
Can you elaborate on the exit criterias, please?

With mutable state and error we can have a different method signature and simplified flow:

```go
Accept(msg sdk.Msg, block abci.Header) (persist bool, err error)
```

Reply:
Here's the current implementation: https://github.com/cosmos-gaians/cosmos-sdk/blob/hackatom/x/delegation/keeper.go#L83-L112

As a functional programmer, my instinct is to avoid mutable state. But for a blockchain, I would defer to whatever is more efficient. Do we know the capability can actually be mutated? i.e. do we know we're getting a pointer not a value reference?